pkg/api/licenses.go

```diff
@@ -720,41 +721,78 @@ func addChangelogsForLicenseUpdate(tx *gorm.DB, username string,
-		((oldLicense.ExternalRef.Data().InternalComment != nil && newLicense.ExternalRef.Data().InternalComment != nil) && (*oldLicense.ExternalRef.Data().InternalComment != *newLicense.ExternalRef.Data().InternalComment)) {
-		changes = append(changes, models.ChangeLog{
-			Field:        "ExternalRef.InternalComment",
-			OldValue:     oldLicense.ExternalRef.Data().InternalComment,
-			UpdatedValue: newLicense.ExternalRef.Data().InternalComment,
-		})
-	}
-	if (oldLicense.ExternalRef.Data().LicenseExplanation == nil && newLicense.ExternalRef.Data().LicenseExplanation != nil) ||
-		(oldLicense.ExternalRef.Data().LicenseExplanation != nil && newLicense.ExternalRef.Data().LicenseExplanation == nil) ||
-		((oldLicense.ExternalRef.Data().LicenseExplanation != nil && newLicense.ExternalRef.Data().LicenseExplanation != nil) && (*oldLicense.ExternalRef.Data().LicenseExplanation != *newLicense.ExternalRef.Data().LicenseExplanation)) {
-		changes = append(changes, models.ChangeLog{
-			Field:        "ExternalRef.LicenseExplanation",
-			OldValue:     oldLicense.ExternalRef.Data().LicenseExplanation,
-			UpdatedValue: newLicense.ExternalRef.Data().LicenseExplanation,
-		})
-	}
-	if (oldLicense.ExternalRef.Data().LicenseSuffix == nil && newLicense.ExternalRef.Data().LicenseSuffix != nil) ||
-		(oldLicense.ExternalRef.Data().LicenseSuffix != nil && newLicense.ExternalRef.Data().LicenseSuffix == nil) ||
-		((oldLicense.ExternalRef.Data().LicenseSuffix != nil && newLicense.ExternalRef.Data().LicenseSuffix != nil) && (*oldLicense.ExternalRef.Data().LicenseSuffix != *newLicense.ExternalRef.Data().LicenseSuffix)) {
-		changes = append(changes, models.ChangeLog{
-			Field:        "ExternalRef.LicenseSuffix",
-			OldValue:     oldLicense.ExternalRef.Data().LicenseSuffix,
-			UpdatedValue: newLicense.ExternalRef.Data().LicenseSuffix,
-		})
-	}
-	if (oldLicense.ExternalRef.Data().LicenseVersion == nil && newLicense.ExternalRef.Data().LicenseVersion != nil) ||
-		(oldLicense.ExternalRef.Data().LicenseVersion != nil && newLicense.ExternalRef.Data().LicenseVersion == nil) ||
-		((oldLicense.ExternalRef.Data().LicenseVersion != nil && newLicense.ExternalRef.Data().LicenseVersion != nil) && (*oldLicense.ExternalRef.Data().LicenseVersion != *newLicense.ExternalRef.Data().LicenseVersion)) {
-		changes = append(changes, models.ChangeLog{
-			Field:        "ExternalRef.LicenseVersion",
-			OldValue:     oldLicense.ExternalRef.Data().LicenseVersion,
-			UpdatedValue: newLicense.ExternalRef.Data().LicenseVersion,
-		})
+
+	oldLicenseExternalRef := oldLicense.ExternalRef.Data()
+	oldExternalRefVal := reflect.ValueOf(oldLicenseExternalRef)
+	typesOf := oldExternalRefVal.Type()
+
+	newLicenseExternalRef := newLicense.ExternalRef.Data()
+	newExternalRefVal := reflect.ValueOf(newLicenseExternalRef)
+
+	for i := 0; i < oldExternalRefVal.NumField(); i++ {
+		fieldName := typesOf.Field(i).Name
+
+		switch typesOf.Field(i).Type.String() {
```

Comment:
These datatype checks will be required at CreateLicense as well?

Reply:
These checks are only needed for adding fields to the changelog. There was checking previously too, to avoid nil pointer dereference while converting bool, int etc to string pointer for updating the changelog values. As the fields are not known at the time of coding and generated at compile time, this generic approach had to be taken here.

Reply:
Ah, that makes sense, changelog holds a generic string pointer for previous and new values.